### PR TITLE
Rework of the batching mechanism to allow for more accurate effective size of the batch requests

### PR DESCRIPTION
**Pull request by Horneth**

### CODE REVIEW

File: core/src/main/scala/cromwell/core/actor/BatchActor.scala [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  /** The states for BatchActor. */
+  sealed trait BatchActorState
+  case object WaitingToProcess extends BatchActorState
+  case object Processing extends BatchActorState
+
+  /** The control messages for BatchActor.
+    *  Because they extends ControlMessage, in combination with a control aware mailbox those messages will
+    *  be processed in priority.
+    */
+  sealed trait BatchActorControleMessage extends ControlMessage
```
Comment (danbills) [inline core/src/main/scala/cromwell/core/actor/BatchActor.scala]: *Control

File: core/src/main/resources/reference.conf [hunk 2]
```diff
@@ -23,6 +23,10 @@ akka {
     #parallelism-max = 64
   }
 
+  priority-mailbox {
+    mailbox-type = "akka.dispatch.UnboundedControlAwareMailbox"
```
Comment (mcovarr) [inline core/src/main/resources/reference.conf]: 👁

File: core/src/main/scala/cromwell/core/actor/BatchActor.scala [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+  def commandToData(snd: ActorRef): PartialFunction[Any, C]
+  
+  when(WaitingToProcess) {
+    // On a regular event, only process if we're above the batch size is reached
+    case Event(command, data) if commandToData(sender).isDefinedAt(command) =>
+      processIfBatchSizedReached(data.enqueue(commandToData(sender)(command)))
+    // On a scheduled process, always process. Use the opportunity to broadcast the current queue weight
+    case Event(ScheduledProcessAction, data) =>
+      gossip(QueueWeight(data.weight))
```
Comment (mcovarr) [inline core/src/main/scala/cromwell/core/actor/BatchActor.scala]: 😮 😎

Comment (mcovarr) [reply core/src/main/scala/cromwell/core/actor/BatchActor.scala]: does this data make me look fat?

File: core/src/main/scala/cromwell/core/actor/BatchActor.scala [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    case Event(ShutdownCommand, _) =>
+      shuttingDown = true
+      stay()
+  }
+  
+  protected def sizeFunction(command: C): Int
+  
+  protected def process(data: Vector[C]): Future[Int]
+  
+  private def processIfBatchSizedReached(data: BatchData[C]) = {
```
Comment (mcovarr) [inline core/src/main/scala/cromwell/core/actor/BatchActor.scala]: Sized => Size

File: core/src/test/scala/cromwell/core/actor/BatchActorSpec.scala [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  it should "go back to WaitingToProcess after processing if not over batch size" in {
+    // Add some processing time so we have time to send more events while a batch is being processed
+    val batch = TestFSMRef(new BatchActorTest(2.seconds))
+    batch ! "hola"
+    batch ! "hello"
+    batch ! "bonjour"
+
+    // at tha point we've processed the first batch and are waiting for the future to complete
```
Comment (mcovarr) [inline core/src/test/scala/cromwell/core/actor/BatchActorSpec.scala]: tha point => the point, unless this is referring to the Biggie track

File: core/src/test/scala/cromwell/core/actor/BatchActorSpec.scala [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  it should "process again when previous processing finished and we're still over batch size" in {
+    // Add some processing time so we have time to send more events while a batch is being processed
+    val batch = TestFSMRef(new BatchActorTest(2.seconds))
+    batch ! "hola"
+    batch ! "hello"
+    batch ! "bonjour"
+
+    // at tha point we've processed the first batch and are waiting for the future to complete
```
Comment (mcovarr) [inline core/src/test/scala/cromwell/core/actor/BatchActorSpec.scala]: chorus

File: core/src/main/scala/cromwell/core/actor/BatchActor.scala [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    case Event(ProcessingComplete, data) if !shuttingDown =>
+      processIfBatchSizedReached(data)
+    case Event(ShutdownCommand, _) =>
+      shuttingDown = true
+      stay()
+  }
+  
+  protected def sizeFunction(command: C): Int
+  
+  protected def process(data: Vector[C]): Future[Int]
```
Comment (mcovarr) [inline core/src/main/scala/cromwell/core/actor/BatchActor.scala]: doc the meaning of the returned `Future[Int]` plz

File: services/src/main/scala/cromwell/services/instrumentation/InstrumentedBatchActor.scala [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+object InstrumentedBatchActor {
+  case object QueueSizeTimerKey  
+  case object QueueSizeTimerAction  
+}
+
+/**
+  * Layer over batch actor than instruments the throughput and queue size
```
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/instrumentation/InstrumentedBatchActor.scala]: than => that

File: core/src/main/scala/cromwell/core/actor/BatchActor.scala (hunk 3 above)
Comment (geoffjentry) [reply core/src/main/scala/cromwell/core/actor/BatchActor.scala]: I'd only ever seen gossip in akka as being tied to cluster stuff. what is this?

File: common/src/main/scala/common/collections/EnhancedCollections.scala [hunk 9]
```diff
@@ -25,4 +27,59 @@ object EnhancedCollections {
...
+      * The implementation is simple but naive. The more general problem is described here https://en.wikipedia.org/wiki/Knapsack_problem
+      * 
+      * @tparam W a numeric type for the weight
+      * @return (head, newQueue)
+      */
+    def takeWhileWeighted[W](maxWeight: W,
+                             weightFunction: A => W,
+                             maxHeadLength: Option[Int],
+                             strict: Boolean = false)
+                            (implicit n: Numeric[W], c: Ordering[W]): (Vector[A], Queue[A]) = {
```
[the remark this answers is not in the document]
Comment (Horneth) [reply common/src/main/scala/common/collections/EnhancedCollections.scala]: The scala `dequeue` method for `Queue`s returns Tuple so I followed the same movement but I can make a case class for this one

File: common/src/main/scala/common/collections/EnhancedCollections.scala [hunk 10]
```diff
@@ -25,4 +27,59 @@ object EnhancedCollections {
...
+      import n._
```
[the remark this answers is not in the document]
Comment (Horneth) [reply common/src/main/scala/common/collections/EnhancedCollections.scala]: yeah this is so you can use `+`, `-`, etc.. on `Numeric[W]`

File: core/src/main/resources/reference.conf (hunk 2 above)
Comment (Horneth) [reply core/src/main/resources/reference.conf]: This allows messages that extend the `ControlMessage` trait to be processed immediately. I'll add a comment here

File: core/src/main/scala/cromwell/core/actor/BatchActor.scala (hunk 3 above)
Comment (Horneth) [reply core/src/main/scala/cromwell/core/actor/BatchActor.scala]: it's a built-in way simple eay to do event listening. You send a `Listen(your_actor)` to an FSM (or any actor extending the `Listeners` trait). It remembers you and when you call `gossip(message)` it sends the message to all registered listeners.

File: common/src/main/scala/common/collections/EnhancedCollections.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (Horneth) [reply common/src/main/scala/common/collections/EnhancedCollections.scala]: Actually now that I go through this it's pretty handy to be able to just do `val (head, tail) = queue.takeWhileWeighted(...)` without having to unpack a case class
I'll add better scala doc

[the remark this answers is not in the document]
Comment (mcovarr) [reply common/src/main/scala/common/collections/EnhancedCollections.scala]: mokay though case classes do have a built in unapply

```val QueueThing(head, tail) = ...```

**Pull request merged**